src/lib/router.ts

```diff
@@ -74,119 +67,190 @@ const allRoutes: RouteDefinition[] = [
 
-  activePath: String
-  routes: Route[]
-  rootRouter: Router
+    this.routes = routes ? routes : []
 
-  constructor(routes?: Route[], rootRouter?: Router) {
-    if (routes) {
-      this.routes = routes
-    } else {
-      this.routes = transformRoutes(allRoutes)
+    if (parent) {
+      this.parent = parent
     }
+  }
+
+  get fullName(): string {
+    return [this.parent && this.parent.fullName, this.name]
+      .filter(part => part && part !== "")
+      .join(".")
+  }
+
+  get fullPath(): string {
+    return [this.parent && this.parent.fullPath, this.path]
+      .filter(part => part && part !== "")
+      .join("")
+  }
+
+  get parent(): Route {
+    return this._parent
+  }
 
-    if (rootRouter) {
-      this.rootRouter = rootRouter
+  set parent(parent: Route) {
+    this._parent = parent
+    if (!parent.routes.includes(this)) {
+      parent.routes.push(this)
     }
   }
 
-  // Flatten all routes
-  get flattenedRoutes(): Routes {
-    if (!this._flattenedRoutes) {
-      function flatten(routes: Route[]) {
-        return routes.reduce((flattenedRoutes, { routes, ...rest }) => {
-          return [
-            ...flattenedRoutes,
-            ...(routes.length > 0 ? flatten(routes) : rest),
-          ]
-        }, [])
+  get routes(): Route[] {
+    return this._routes
+  }
+
+  set routes(routes: Route[]) {
+    // if we're ever going to remove routes we should address that here
+    this._routes = routes
+    routes.forEach(route => {
+      if (route.parent && route.parent !== this) {
+        throw new Error(
+          `Cannot add ${route.fullName} to ${this.fullName}, because it already belongs to ${route.parent.fullName}`
+        )
       }
+      route.parent = this
+    })
+  }
 
-      this._flattenedRoutes = flatten(this.routes)
+  get allRoutes(): Route[] {
+    let flatten = function(routes: Route[]) {
+      return routes.reduce((result, route) => {
+        return [...result, ...[route], ...flatten(route.routes)]
+      }, [])
     }
 
-    return this._flattenedRoutes
+    return flatten(this.routes)
   }
 
-  // Return the active route
-  get activeRoute(): Route {
-    if (this.rootRouter) {
-      return this.rootRouter.activeRoute
+  get pages(): Route[] {
+    return this.allRoutes.filter(route => route.routes.length === 0)
+  }
+
+  get activePath(): string {
+    return this.parent ? this.parent.activePath : this._activePath
+  }
+
+  set activePath(path: string) {
+    if (this.parent) {
+      throw "activePath can only be set on the router, not a child route"
     } else {
-      return this.flattenedRoutes.find(route => {
-        return route.fullPath.match(this.activePath.replace(/\/+$/, ""))
-      })
+      this._activePath = path
     }
   }
 
-  // Return the previous route
-  get previousRoute(): Route {
-    let match = this.flattenedRoutes.find(
-      route => route.fullName === this.activeRoute.fullName
-    )
+  // Return the active route
+  get activeRoute(): Route {
```

Comment:
maybe `activePage`? and `nextPage` and `previousPage`?

Reply:
basically there is only ever one active page. Every route that's a `parent` is an active route.

Reply:
👍